SPEC.md

```diff
@@ -87,33 +88,58 @@ Network configuration in JSON format is streamed to the plugin through stdin. Th
 
 ### Result
 
-Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details).
+Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. The `ip` and `dns` items should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details) except that the plugin should fill in the `interface` indexes appropriately, which are missing from IPAM plugin output since IPAM plugins should be unaware of interfaces.
 
 ```
 {
-  "cniVersion": "0.2.0",
-  "ip4": {
-    "ip": <ipv4-and-subnet-in-CIDR>,
-    "gateway": <ipv4-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv4-routes>    (optional)
-  },
-  "ip6": {
-    "ip": <ipv6-and-subnet-in-CIDR>,
-    "gateway": <ipv6-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv6-routes>    (optional)
-  },
+  "cniVersion": "0.3.0",
+  "interfaces": [
+      {
+          "name": "<name>",
+          "mac": "<MAC address>",                            (required if L2 addresses are meaningful)
+          "sandbox": "<netns path or hypervisor identifier>" (required for container/hypervisor interfaces, empty/omitted for host interfaces)
+      }
+  ],
+  "ip": [
+      {
+          "version": "<4-or-6>",
+          "address": "<ip-and-prefix-in-CIDR>",
+          "gateway": "<ip-address-of-the-gateway>",      (optional)
+          "interface": "<index into 'interfaces' list>",
+          "routes": [                                    (optional)
```

Comment:
Per discussion with CNI maintainers, we agreed to move Routes out of the IPs structure for now, but leave GW in each address since that's harder to extract at this time due to the bridge plugin.  We also have not made Routes.Gw a list yet, but only to get this large PR merged.

Following this PR I will do another PR (or if I don't, certainly entertain one!) to do all or some of:
1) rename Routes.Gw -> Routes.NextHops
2) make Routes.NextHops an array
3) make Routes.NextHops mandatory
4) remove the "Gateway" field from the IPConfig structure

All these things would require copying types.Route out of pkg/types/types.go into *both* pkg/types/current and pkg/types/020, then modifying the pkg/types/current/types.go version with the required changes.